src/btreemap/node/v2.rs

```diff
@@ -0,0 +1,402 @@
+
+        Self {
+            address,
+            keys,
+            encoded_values: RefCell::new(encoded_values),
+            children,
+            node_type,
+            version: Version::V2(page_size),
+            overflow: if original_overflow_address == crate::types::NULL {
+                None
+            } else {
+                Some(original_overflow_address)
+            },
+        }
+    }
+
+    // Saves the node to memory.
+    pub(super) fn save_v2<M: Memory>(&self, allocator: &mut Allocator<M>) {
+        assert_eq!(self.keys.len(), self.encoded_values.borrow().len());
+
+        let page_size = self.version.page_size();
+
+        // A buffer to serialize the node into first, then write to memory.
+        let mut buf = vec![];
+        buf.extend_from_slice(MAGIC);
+        buf.extend_from_slice(&[LAYOUT_VERSION_2]);
+        buf.push(match self.node_type {
+            NodeType::Leaf => LEAF_NODE_TYPE,
+            NodeType::Internal => INTERNAL_NODE_TYPE,
+        });
+        buf.extend_from_slice(&(self.keys.len() as u16).to_le_bytes());
+
+        // Add a null overflow address. This might get overwritten later in case the node
+        // does overflow.
+        buf.extend_from_slice(&[0; 8]);
+
+        // Load all the values. This is necessary so that we don't overwrite referenced
+        // values when writing the entries to the node.
+        let memory = allocator.memory();
+        for i in 0..self.keys.len() {
+            self.value(i, memory);
+        }
+
+        // Write the children
+        for child in self.children.iter() {
+            buf.extend_from_slice(&child.get().to_le_bytes());
+        }
+
+        // Write the keys.
+        for key in self.keys.iter() {
+            // Write the size of the key.
+            let key_bytes = key.to_bytes();
+            buf.extend_from_slice(&(key_bytes.len() as u32).to_le_bytes());
+
+            // Write the key.
+            buf.extend_from_slice(key_bytes.borrow());
+        }
+
+        // Write the values.
+        for idx in 0..self.entries_len() {
+            // Write the size of the value.
+            let value = self.value(idx, memory);
+            buf.extend_from_slice(&(value.len() as u32).to_le_bytes());
+
+            // Write the value.
+            buf.extend_from_slice(&value);
+        }
+
+        self.write_paginated(buf, allocator, page_size as usize);
+    }
+
+    // Writes a buffer into pages of the given page size.
+    // Pages can be allocated and deallocated as needed.
+    fn write_paginated<M: Memory>(
+        &self,
+        buf: Vec<u8>,
+        allocator: &mut Allocator<M>,
+        page_size: usize,
+    ) {
+        // Compute how many overflow pages are needed.
+        let additional_pages_needed = if buf.len() > page_size {
+            let overflow_page_capacity = page_size - PAGE_OVERFLOW_DATA_OFFSET.get() as usize;
```

Comment:
The subtraction here I suppose assumes that `page_size` > PAGE_OVERFLOW_DATA_OFFSET which should be true indeed. Perhaps we throw an `assert!` here about this condition just to be extra careful?

Reply:
If not mistaken the assertion you mention is `page_size >= MINIMUM_PAGE_SIZE`. I suppose this should be enough given that `MINIMUM_PAGE_SIZE` > `PAGE_OVERFLOW_DATA_OFFSET` but was thinking that an explicit `page_size >= PAGE_OVERFLOW_DATA_OFFSET` doesn't hurt so we might as well add it.

Reply:
Sure, done.